internal/controllers/gateway/route_utils.go

```diff
@@ -313,20 +318,34 @@ func listenerHostnameIntersectWithRouteHostnames[H types.HostnameT, L types.List
 	return false
 }
 
+// isListenterHostnameEffective if the listener could specify an effective hostname to match hostnames in requests.
+// it will return true when the protocol of listener is HTTP, HTTPS, or TLS.
+func isListenerHostnameEffective(listener gatewayv1beta1.Listener) bool {
+	return listener.Protocol == gatewayv1beta1.HTTPProtocolType ||
+		listener.Protocol == gatewayv1beta1.HTTPSProtocolType ||
+		listener.Protocol == gatewayv1beta1.TLSProtocolType
+}
+
 // filterHostnames accepts a HTTPRoute and returns a version of the same object with only a subset of the
 // hostnames, the ones matching with the listeners' hostname.
-func filterHostnames(gateways []supportedGatewayWithCondition, httpRoute *gatewayv1beta1.HTTPRoute) *gatewayv1beta1.HTTPRoute {
+// it returns an error if the intersection of hostname match in httproute and listeners is empty.
+func filterHostnames(gateways []supportedGatewayWithCondition, httpRoute *gatewayv1beta1.HTTPRoute) (*gatewayv1beta1.HTTPRoute, error) {
 	filteredHostnames := make([]gatewayv1beta1.Hostname, 0)
 
 	// if no hostnames are specified in the route spec, get all the hostnames from
 	// the gateway
 	if len(httpRoute.Spec.Hostnames) == 0 {
 		for _, gateway := range gateways {
 			for _, listener := range gateway.gateway.Spec.Listeners {
-				if listenerName := gateway.listenerName; listenerName == "" || listenerName == string(listener.Name) {
-					if listener.Hostname != nil {
-						filteredHostnames = append(filteredHostnames, (*listener.Hostname))
+				listenerName := gateway.listenerName
+				if (listenerName == "" || listenerName == string(listener.Name)) &&
+					isListenerHostnameEffective(listener) {
+					// httpRoute remains an empty hostname if any of the listeners
+					// has not specified hostname to match any host.
+					if listener.Hostname == nil {
+						return httpRoute, nil
```

Comment:
I don't understand why do we want to return here. There's still more listeners and gateways to look it, is it not?

Reply:
In this branch, `hostnames` in `HTTPRoute` is empty to specify that the `HTTPRoute` could match any hostname. So, if any listeners could be attached to this `HTTPRoute` has an empty hostname, that could also match any hostname, the final result of hostname match should be match any hostname. If we finally got a listener with a non-empty hostname, a hostname will be included in the `HTTPRoute`, so that the `HTTPRoute` will be only match certain hostnames, and this is not expected: #3175